application/main.py

```diff
@@ -573,7 +572,92 @@ def __eq__(self, other):
                (self.validation_results == other.validation_results)
 
 
+@application.route('/build_implementer_agreement_results/<id>')
+@login_required
+def build_implementer_agreement_results(user_data, id):
+    with database.Session() as session:
+
+        model = session.query(database.model).filter(database.model.code == id).all()[0]
+        tasks = {t.task_type: t for t in model.tasks}
+        instances = {i: i for i in model.instances}
+
+        imp_agreement_results = [imp_ag.serialize() for imp_ag in tasks['implementer_agreements_task'].results]
+        instances_results = [ins.serialize() for ins in instances]
+
+        then_statements = []
+        unique_steps = set()
+
+        for imp_ag in tasks['implementer_agreements_task'].results:
+            imp_ag = imp_ag.serialize()
+            unique_steps.add(imp_ag['step'])
+        for unique_step in unique_steps:
+            results_related_to_step = [result for result in imp_agreement_results if result['step'] == unique_step]
+            imp_ag_dict = {}
+            imp_ag_dict['text'] = unique_step
+            imp_ag_dict['lineno'] = None
+            step_results = []
+            for result in results_related_to_step:
+                step_results_dict = {}
+                instances_row = next((item for item in instances_results if item["id"] == result['instance_id']), None)
+                step_results_dict['entity_id'] = result.get('instance_id')
+                step_results_dict['msg'] = result.get('message')
+                if result['message'] == 'Rule passed':
+                    step_results_dict['status'] = 'v'
+                elif result['message'] == 'Rule disabled':
+                    step_results_dict['status'] = 'n'
+                else:
+                    step_results_dict['status'] = 'i'
+                step_results_dict['entity_type'] = instances_row.get('ifc_type')
+                step_results.append(step_results_dict)
+            imp_ag_dict['results'] = step_results
+            then_statements.append(imp_ag_dict)
+
+    return jsonify(then_statements)
+
+@application.route('/build_informal_propositions_results/<id>')
+@login_required
+def build_informal_propositions_results(user_data, id):
+    with database.Session() as session:
 
+        model = session.query(database.model).filter(database.model.code == id).all()[0]
+        tasks = {t.task_type: t for t in model.tasks}
+        instances = {i: i for i in model.instances}
+
+        inf_propositions_results = [imp_ag.serialize() for imp_ag in tasks['informal_propositions_task'].results]
+        instances_results = [ins.serialize() for ins in instances]
+
+        then_statements = []
+        unique_steps = set()
```

Comment:
Same here 
`unique_steps = set(result['step'] for result in inf_propositions_result)`

Reply:
Done.